src/include/plan_node/output_schema.h

```diff
@@ -0,0 +1,228 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include "common/constants.h"
+#include "common/hash_util.h"
+#include "common/macros.h"
+#include "common/strong_typedef.h"
+#include "parser/expression/abstract_expression.h"
+#include "plan_node/plan_node_defs.h"
+#include "storage/storage_defs.h"
+#include "type/type_id.h"
+#include "type/type_util.h"
+
+namespace terrier::plan_node {
+
+/**
+ * Internal object for representing output columns of a plan node. This object is to be differentiated from
+ * catalog::Schema, which contains all columns of a table. This object can contain partial
+ */
+class OutputSchema {
+ public:
+  /**
+   * This object contains output columns of a plan node which can consist of columns that exist in the catalog
+   * or intermediate columns.
+   */
+  class Column {
+   public:
+    /**
+     * Instantiates a Column object, primary to be used for building a Schema object
+     * @param name column name
+     * @param type SQL type for this column
+     * @param oid internal unique identifier for this column
+     */
+    Column(std::string name, const type::TypeId type, const bool nullable, const catalog::col_oid_t oid)
+        : name_(std::move(name)), type_(type), nullable_(nullable), oid_(oid) {
+      TERRIER_ASSERT(type_ != type::TypeId::INVALID, "Attribute type cannot be INVALID.");
+    }
+    /**
+     * @return column name
+     */
+    const std::string &GetName() const { return name_; }
+    /**
+     * @return SQL type for this column
+     */
+    type::TypeId GetType() const { return type_; }
+    /**
+     * @return true if the column is nullable, false otherwise
+     */
+    bool GetNullable() const { return nullable_; }
+    /**
+     * @return internal unique identifier for this column
+     */
+    catalog::col_oid_t GetOid() const { return oid_; }
+    /**
+     * @return the hashed value for this column based on name and OID
+     */
+    common::hash_t Hash() const {
+      common::hash_t hash = common::HashUtil::Hash(name_);
+      hash = common::HashUtil::CombineHashes(hash, common::HashUtil::Hash(oid_));
+      return hash;
+    }
+    /**
+     * @return whether the two columns are equal
+     */
+    bool operator==(const Column &rhs) const { return name_ == rhs.name_ && type_ == rhs.type_ && oid_ == rhs.oid_; }
+    /**
+     * Inequality check
+     * @param rhs other
+     * @return true if the two columns are not equal
+     */
+    bool operator!=(const Column &rhs) const { return !operator==(rhs); }
+
+   private:
+    const std::string name_;
+    const type::TypeId type_;
+    const bool nullable_;
+    const catalog::col_oid_t oid_;
+  };
+
+  /**
+   * An intermediate column produced by plan nodes
+   */
+  struct DerivedColumn {
```

Comment:
Why is `OutputSchema::Column` a class but `OutputSchema::DerivedColumn` is a struct? I think we should be consistent.